### PR TITLE
enforce compatible es version when deploying to heroku

### LINKED ISSUE #18650: App with Elasticsearch doesn't work out-of-the-box with Heroku
##### **Overview of the issue**

I'm updating 21-Points Health as part of the JHipster Mini-Book. When I deploy to Heroku, it fails to configure the Elasticsearch add-on. I experienced a [similar issue](https://github.com/jhipster/generator-jhipster/issues/7810) when I updated the book for JHipster v5.

When I run `jhipster heroku`, it does mention something about Elasticsearch in that "no new Elasticsearch addon created".

```
Provisioning addons

Provisioning database addon heroku-postgresql --as DATABASE

No suitable cache addon for cacheprovider ehcache available.

Creating Heroku deployment files
    force .yo-rc-global.json
   create Procfile
    force .yo-rc.json
   create system.properties
   create gradle/heroku.gradle
 conflict build.gradle
? Overwrite build.gradle? (ynarxdeiH) No new Elasticsearch addon created
? Overwrite build.gradle? overwrite this and all others
```

When I tail the logs after startup, it basically says the [Bonsai Elasticsearch add-on](https://elements.heroku.com/addons/bonsai) wasn't configured.

```
2022-05-13T15:31:15.077985+00:00 app[web.1]: Caused by: java.net.UnknownHostException: 
  ${BONSAI_URL}: Name or service not known
```

I tried installing the add-on manually:

```
$ heroku addons:create bonsai:sandbox-6
Creating bonsai:sandbox-6 on ⬢ health-by-points... free
Your Bonsai cluster is being provisioned.

Please run `heroku addons:open bonsai` and navigate to the Notifications tab to enter contact information for this resource.

Please contact support@bonsai.io for help.
bonsai-clean-03403 is being created in the background. The app will restart when complete...
Use heroku addons:info bonsai-clean-03403 to check creation progress
Use heroku addons:docs bonsai to view documentation
```

Then, I ran `heroku restart`. Now it fails with:

```
... app[web.1]: Caused by: org.elasticsearch.ElasticsearchException: Invalid or missing build flavor [oss]
... app[web.1]: at org.elasticsearch.client.RestHighLevelClient.performClientRequest(RestHighLevelClient.java:2084)
... app[web.1]: at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1732)
... app[web.1]: at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1717)
... app[web.1]: at org.elasticsearch.client.IndicesClient.exists(IndicesClient.java:963)
... app[web.1]: at org.springframework.data.elasticsearch.core.RestIndexTemplate.lambda$doExists$2(RestIndexTemplate.java:103)
... app[web.1]: at org.springframework.data.elasticsearch.core.ElasticsearchRestTemplate.execute(ElasticsearchRestTemplate.java:408)
```

If I google the error, I land at https://github.com/elastic/elasticsearch/issues/76091. 

##### **Motivation for or Use Case**

Elasticsearch + Heroku should work out-of-the-box. 

##### **Reproduce the error**

Create a new app with Elasticsearch and deploy it to Heroku with `jhipster heroku`. 

##### **Related issues**

We experienced a similar issue in June 2018: https://github.com/jhipster/generator-jhipster/issues/7810

##### **Suggest a Fix**

Fix the Heroku sub-generator so it works.

##### **JHipster Version(s)**

```
twenty-one-points@0.0.1-SNAPSHOT /Users/mraible/dev/21-points
└── generator-jhipster@7.8.1

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "applicationType": "monolith",
    "authenticationType": "jwt",
    "baseName": "TwentyOnePoints",
    "blueprints": [],
    "buildTool": "gradle",
    "cacheProvider": "ehcache",
    "clientFramework": "angularX",
    "clientPackageManager": "npm",
    "clientTheme": "none",
    "clientThemeVariant": "",
    "creationTimestamp": 1651460089159,
    "cypressCoverage": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "devServerPort": 4200,
    "dtoSuffix": "DTO",
    "enableGradleEnterprise": false,
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "entitySuffix": "",
    "herokuAppName": "health-by-pints",
    "herokuDeployType": "git",
    "herokuJavaVersion": "11",
    "jhiPrefix": "jhi",
    "jhipsterVersion": "7.8.1",
    "jwtSecretKey": "YourJWTSecretKeyWasReplacedByThisMeaninglessTextByTheJHipsterInfoCommandForObviousSecurityReasons",
    "languages": ["en", "fr"],
    "messageBroker": false,
    "nativeLanguage": "en",
    "otherModules": [],
    "packageName": "org.jhipster.health",
    "pages": [],
    "prodDatabaseType": "postgresql",
    "reactive": false,
    "searchEngine": "elasticsearch",
    "serverPort": "8080",
    "serverSideOptions": ["searchEngine:elasticsearch"],
    "serviceDiscoveryType": "no",
    "skipCheckLengthOfIdentifier": false,
    "skipFakeData": false,
    "skipUserManagement": false,
    "testFrameworks": ["cypress"],
    "websocket": false,
    "withAdminUi": true
  }
}

</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>

</pre>
</details>

##### **Environment and Tools**

openjdk version "17.0.3" 2022-04-19
OpenJDK Runtime Environment GraalVM CE 22.1.0 (build 17.0.3+7-jvmci-22.1-b06)
OpenJDK 64-Bit Server VM GraalVM CE 22.1.0 (build 17.0.3+7-jvmci-22.1-b06, mixed mode, sharing)

git version 2.32.0 (Apple Git-132)

node: v16.15.0

npm: 8.6.0

Docker version 20.10.14, build a224086

Docker Compose version v2.5.0

##### **Browsers and Operating System**

Chrome and macOS Monterey

- [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by atomfrede**
- [x] enforce dependency versions via maven

closes #18650

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: generators/heroku/index.js [hunk 1]
```diff
@@ -489,13 +489,15 @@ module.exports = class extends BaseBlueprintGenerator {
 
         this.log(chalk.bold('\nProvisioning addons'));
         if (this.searchEngine === ELASTICSEARCH) {
+          this.log(chalk.bold('\nProvisioning bonsai elasticsearch addon'));
           ChildProcess.exec(
-            `heroku addons:create bonsai:sandbox --as BONSAI --app ${this.herokuAppName}`,
-            addonCreateCallback.bind(this, 'Elasticsearch')
-          );
+            `heroku addons:create bonsai:sandbox-6 --as BONSAI --app ${this.herokuAppName}`, (err, stdout, stderr) => {
```
Comment (atomfrede) [inline generators/heroku/index.js:494]: Important is `-6`. Just using `sandbox` does not work as the tier does not exist.

File: generators/heroku/templates/application-heroku.yml.ejs [hunk 2]
```diff
@@ -76,8 +76,7 @@ spring:
 <%_ } _%>
 <%_ if (searchEngineElasticsearch) { _%>
   elasticsearch:
-    rest:
-      uris: ${BONSAI_URL}
+    uris: ${BONSAI_URL}
```
Comment (atomfrede) [inline generators/heroku/templates/application-heroku.yml.ejs:79]: Remove deprecated property usage, aligned with prod/dev properties

**Comment by atomfrede** [comment]
Sorry, forgot to adapt the snapshots for changed properties file 🙈 🤦‍♂️

**Comment by atomfrede** [comment]
Updated the documentation too: https://github.com/jhipster/jhipster.github.io/pull/1214

File: generators/heroku/templates/heroku.gradle.ejs [hunk 3]
```diff
@@ -22,8 +22,26 @@ heroku {
...
+
 <%_ if (herokuDeployType === 'git') { _%>
 
+if (System.getenv("DYNO") != null) {
+    configurations {
+        all {
+            resolutionStrategy {
+                // force dependency version as used bonsai add-on as of now only supports 7.10.x
+                // https://github.com/jhipster/generator-jhipster/issues/18650
+                force 'org.elasticsearch.client:elasticsearch-rest-client:7.10.2'
```
Comment (Tcharl) [inline generators/heroku/templates/heroku.gradle.ejs:36]: version in gradle-property?

File: generators/heroku/templates/heroku.gradle.ejs [hunk 4]
```diff
@@ -38,4 +56,17 @@ gradle.taskGraph.whenReady {taskGraph ->
...
 }
-<%_ } _%>
+<%_ } else { _%>
+configurations {
+    all {
+        resolutionStrategy {
+            // force dependency version as used bonsai add-on as of now only supports 7.10.x
+            // https://github.com/jhipster/generator-jhipster/issues/18650
+            force 'org.elasticsearch.client:elasticsearch-rest-client:7.10.2'
+            force 'org.elasticsearch.client:elasticsearch-rest-high-level-client:7.10.2'
```
Comment (Tcharl) [inline generators/heroku/templates/heroku.gradle.ejs:66]: versions in gradle-property?

File: generators/heroku/templates/pom-profile.xml.ejs [hunk 5]
```diff
@@ -1,3 +1,27 @@
+            <dependencyManagement>
+                <dependencies>
+                    <dependency>
+                        <groupId>org.elasticsearch.client</groupId>
+                        <artifactId>elasticsearch-rest-client</artifactId>
+                        <version>7.10.2</version>
```
Comment (Tcharl) [inline generators/heroku/templates/pom-profile.xml.ejs:6]: version in the In the bom ?
Or in the profile's maven property section?

Comment (atomfrede) [reply generators/heroku/templates/pom-profile.xml.ejs:6]: Sounds good to have it in the profile.

File: generators/heroku/templates/heroku.gradle.ejs (hunk 4 above)
Comment (atomfrede) [reply generators/heroku/templates/heroku.gradle.ejs:66]: Would like to keep it enclosed in the heroku specific file/config, will try to define a property there not in gradle.properties

**Comment by mraible** [comment]
I tested this today by checking out this branch and running `jhipster jdl 21-points.jh`. Then, I ran everything locally with Gradle, Keycloak, and Elasticsearch. `npm run e2e` passes. 

Next, I tried deploying to Heroku with `jhipster heroku` after running `npm link generator-jhipster`. It fails with the following error:

```
Creating Heroku application and setting up node environment
✖ Error: Command failed: heroku create hipsteres
 ›   Warning: heroku update available from 7.59.2 to 7.60.2.
Creating hipsteres...
    at ChildProcess.exithandler (node:child_process:399:12)
    at ChildProcess.emit (node:events:520:28)
    at maybeClose (node:internal/child_process:1092:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5) {
  killed: true,
  code: null,
  signal: 'SIGTERM',
  cmd: 'heroku create hipsteres'
}
    force .yo-rc.json
```

I ran `heroku update` to update the Heroku CLI and ran `git checkout .` to revert the changes to `.yo-rc.json`. Then I tried `jhipster heroku` again. This time, it failed because I'm not logged in.

```
✖ Error: Not authenticated. Run 'heroku login' to login to your heroku account and try again.
    force .yo-rc.json
```

I ran `heroku login`, reverted `.yo-rc.json` and tried again. This time, it worked. However, I didn't have `jq` installed:

```
jq is not available but required. See https://stedolan.github.io/jq/download/ for installation guidance.
```

I installed it with `brew install jq` and ran `./provision-okta-addon.sh` manually. It never seemed to prompt me for a password for the email I entered earlier. That seems to be an issue with Okta + Heroku and not Elasticsearch, so I'll approve this PR.

**Comment by atomfrede** [comment]
bounty claimed https://opencollective.com/generator-jhipster/expenses/80312

**Comment by pascalgrimaud** [comment]
@atomfrede : approved

**Pull request merged**